Add window focus refetching (#13210)
apollographql/apollo-client: .changeset/gentle-windows-watch.md

    sources: { windowFocus: windowFocusSource },
  }),
  defaultOptions: {
    watchQuery: { refetchOn: false },
  },
});

// Only this query refetches on window focus
useQuery(DASHBOARD_QUERY, { refetchOn: { windowFocus: true } });
```

When `defaultOptions.watchQuery.refetchOn` and per-query `refetchOn` options are provided, the objects are merged together.

### Custom events

You can also add your own custom events that trigger refetches. Register your event name and payload type using TypeScript module augmentation, then provide a source function that returns an Observable. The source's emitted value becomes the event's `payload`.

```ts
import { Observable } from "@apollo/client";
import { filter } from "rxjs";
import { AppState, AppStateStatus, Platform } from "react-native";

declare module "@apollo/client" {
  interface RefetchEvents {
    reactNativeAppStatus: AppStateStatus;
  }
}

const refetchEventManager = new RefetchEventManager({
  sources: {
    reactNativeAppStatus: () => {
      return new Observable((observer) => {
        const subscription = AppState.addEventListener("change", observer.next);
        return () => subscription.remove();

>       }).pipe(filter((status) => Platform.OS !== "web" && status === "active"));

Author: Fixed in [043a67a](https://github.com/apollographql/apollo-client/pull/13210/commits/043a67a9922a20a65caf0b97865153392fb59f66)

```ts
declare module "@apollo/client" {
  interface RefetchEvents {
    userTriggered: void;
  }
}

const refetchEventManager = new RefetchEventManager({
  sources: { userTriggered: true },
});

refetchEventManager.emit("userTriggered");
```

Note: Calling `emit` on an event without a registered source will log a warning and result in a no-op.

### Custom handlers

When an event fires, the default handler calls `client.refetchQueries({ include: "active" })` filtered by each query's `refetchOn` setting. You can override the handler for an event to add your own custom filtering. For example, to refetch all queries, including `standby` queries, define a handler for the event:

```ts
const refetchEventManager = new RefetchEventManager({
  // ...
  handlers: {
    userTriggered: ({ client, source, payload, matchesRefetchOn }) => {
      return client.refetchQueries({
        include: "all",
        onQueryUpdated: (observableQuery) => {
          return matchesRefetchOn(observableQuery);
        },
      });
    },
  },
});
```

> Handlers must return a `RefetchQueriesResult`. Conditionally skip a refetch for an event by returning `void`.

Author: `can` feels too relaxed. What about combining sentences? 

```suggestion
Handlers must return either a `RefetchQueriesResult` or `void`. Returning `void` skips refetching for the event.
```

  link,
  cache,
  refetchEventManager: new RefetchEventManager({
    sources: { windowFocus: windowFocusSource },
  }),
  defaultOptions: {
    watchQuery: { refetchOn: false },
  },
});

// Only this query refetches on window focus
useQuery(DASHBOARD_QUERY, { refetchOn: { windowFocus: true } });
```

When `defaultOptions.watchQuery.refetchOn` and per-query `refetchOn` options are provided, the objects are merged together.

### Custom events

You can also add your own custom events that trigger refetches. Register your event name using TypeScript module augmentation, then provide a source for the custom event.

```ts
declare module "@apollo/client" {
  interface RefetchEvents {
    reactNativeAppStatus: true;
  }
}

import { AppState, AppStateStatus } from "react-native";

const refetchEventManager = new RefetchEventManager({
  sources: {
    reactNativeAppStatus: (emit) => {
      function handleChange(status: AppStateStatus) {
        if (Platform.OS !== "web" && status === "active") {
          emit();
        }
      }

      const subscription = AppState.addEventListener("change", handleChange);

      return () => {
        subscription.remove();
      };
    },
  },

> });

Reviewer: Looking at this, should we use `rxjs` more here?

Any source (`windowFocusSource`, `onlineSource`) could be an observable.

```js
import { AppState, AppStateStatus, Platform } from "react-native";
import {Observable, filter} from "rxjs" 

const reactNativeAppStatusSource = new Observable<AppStateStatus>((subscriber) => {
   const subscription = AppState.addEventListener("change", subscriber.next)
   return () => subscription.remove()
}).pipe(filter(status => Platform.OS !== "web" && status === "active"))

const refetchEventManager = new RefetchEventManager({
  sources: {
    reactNativeAppStatus: reactNativeAppStatusSource
  }
});
```

And instead of just `true` in the declaration, we could even have more nuanced types, so
```diff
declare module "@apollo/client" {
  interface RefetchEvents {
-    reactNativeAppStatus: true;
+   reactNativeAppStatus: AppStateStatus
  }
}
```
would be possible.

`RefetchEventManager` itself would only treat emitted events, even with `undefined` as emitted events - but `RefetchEvent` could go from a string to `{ type: "windowFocus" } | { type: "reactNativeAppStatus", value: AppStateStatus }`, which would give the callback even more nuanced situational control.

Author: I suppose you could, but I think the observable only makes sense only if we provide values to next function and do something with that value. Right now that `emit` callback is just a way to tell the manager to start the refetch. I kept it as a simple callback for that reason.

That said, how do you envision the value being used? Are you thinking it would be available through the `refetchOn` callback or something else?

Author: Ok I liked this a lot better. The latest changes include support for observables that emit values based on the `RefetchEvents` value. Thanks for the suggestion!

  link,
  cache: new InMemoryCache(),
  refetchEventManager: new RefetchEventManager({
    sources: {
      // Refetch when window is focused
      windowFocus: windowFocusSource,

      // Refetch when the user comes back online
      online: onlineSource,
    },
  }),
});
```

By default, all active queries refetch when the events fire. Queries can opt out per-event or disable all event refetches:

```ts
// Skip refetch on window focus for this query, but keep `online`
useQuery(QUERY, {
  refetchOn: { windowFocus: false },
});

// Disable all event-driven refetches for this query
useQuery(OTHER_QUERY, {
  refetchOn: false,
});

// Enable every event for this query, regardless of defaultOptions
useQuery(LIVE_DASHBOARD, {
  refetchOn: true,
});
```

To enable per-query opt-in rather than opt-out, set `defaultOptions.watchQuery.refetchOn` to `false` and enable it per-query instead.

```ts
const client = new ApolloClient({
  link,
  cache,
  refetchEventManager: new RefetchEventManager({
    sources: { windowFocus: windowFocusSource },
  }),
  defaultOptions: {
    watchQuery: { refetchOn: false },
  },
});

// Only this query refetches on window focus

> useQuery(DASHBOARD_QUERY, { refetchOn: { windowFocus: true } });

Reviewer: This changelog also needs an example of `refetchOn` with a function.

Author: Updated in [8cf5e21](https://github.com/apollographql/apollo-client/pull/13210/commits/8cf5e21d1105d2c73b2c9fcf62190d2e37e88cf5)
